File: internal/token/token.go
package token

// Kind is the lexical category of a token.
type Kind int

const (
	Unknown Kind = iota
	OpenTag      // <?php <?= <?
	CloseTag     // ?>
	InlineHTML   // text outside PHP tags
	Whitespace   // spaces, tabs, newlines
	Comment      // // # /* */
	DocComment   // /** */
	Variable     // $foo
	Ident        // keywords and names
	Number
	String // '...' "..."
	Punct  // operators, braces, ; , etc.
	EOF
)

var kindNames = map[Kind]string{
	Unknown:    "Unknown",
	OpenTag:    "OpenTag",
	CloseTag:   "CloseTag",
	InlineHTML: "InlineHTML",
	Whitespace: "Whitespace",
	Comment:    "Comment",
	DocComment: "DocComment",
	Variable:   "Variable",
	Ident:      "Ident",
	Number:     "Number",
	String:     "String",
	Punct:      "Punct",
	EOF:        "EOF",
}

func (k Kind) String() string {
	if n, ok := kindNames[k]; ok {
		return n
	}
	return "Kind(?)"
}

// Token is a single lexical unit. The stream is lossless: concatenating every
// Value in order reproduces the original source byte for byte.
type Token struct {
	Kind  Kind
	Value string
	Pos   int // byte offset in source
}

File: rust/src/token.rs
// Token kinds and the Token struct, mirroring ecs-go's internal/token.

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Kind {
    OpenTag,    // <?php <?= <?
    CloseTag,   // ?>
    InlineHtml, // text outside PHP tags
    Whitespace, // spaces, tabs, newlines
    Comment,    // // # /* */
    DocComment, // /** */
    Variable,   // $foo
    Ident,      // names (T_STRING): function/class names, constants, types
    Keyword,    // reserved words
    Number,
    String, // '...' "..."
    Punct,  // operators, braces, ; , etc.
}

// Token value is kept as raw bytes so the stream is byte-for-byte lossless,
// matching Go's use of string (bytes) rather than assuming valid UTF-8.
#[derive(Clone)]
pub struct Token {
    pub kind: Kind,
    pub value: Vec<u8>,
}

impl Token {
    pub fn new(kind: Kind, value: &[u8]) -> Self {
        Token {
            kind,
            value: value.to_vec(),
        }
    }
}
